### PR TITLE
Implementation of Image Search

### PR DESCRIPTION
**Pull request by amawai**
Addresses US #74, and tasks #96, #97, and #98

This feature allows for users to pass a picture or a selected image from their gallery and the application will return a list of relevant labels pertaining to the image. The user may opt to see articles from these labels.

Note that the API used is Clarifai and not Google Images. 

This PR includes the feature code as well as unit and end-to-end tests

### CODE REVIEW

File: app/src/main/java/org/wikipedia/feed/searchbar/SearchCardView.java [hunk 1]
```diff
@@ -1,19 +1,25 @@
 package org.wikipedia.feed.searchbar;
 
 import android.content.Context;
+import android.view.View;
+import android.widget.ImageView;
+import android.widget.Toast;
```
Comment (amawai) [inline app/src/main/java/org/wikipedia/feed/searchbar/SearchCardView.java]: Codacy Issue found: [Unused import - android.widget.Toast.](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541123)

**Comment by amawai** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 96
           

Complexity increasing per file
==============================
- app/src/main/java/org/wikipedia/main/MainFragment.java  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541123&bid=7051231)

File: app/src/main/java/org/wikipedia/main/MainFragment.java [hunk 2]
```diff
@@ -1,16 +1,24 @@
 package org.wikipedia.main;
 
+import android.Manifest;
```
Comment (amawai) [inline app/src/main/java/org/wikipedia/main/MainFragment.java]: Codacy Issue found: [Unused import - android.Manifest.](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541123)

File: app/src/main/java/org/wikipedia/imagesearch/ImageLabeler.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+            conn.setRequestProperty("Authorization", "Key cf037dad49144c56b882dec39ef8a832");
+            conn.setRequestProperty("Content-Type", "application/json");
+            conn.setUseCaches(false);
+            //add json as requestbody
+            OutputStream os = conn.getOutputStream();
+            os.write(inputs.toString().getBytes("UTF-8"));
+            os.close();
+
+            int statusCode = conn.getResponseCode();
+            if (statusCode ==  200) {
```
Comment (amawai) [inline app/src/main/java/org/wikipedia/imagesearch/ImageLabeler.java]: Codacy Issue found: ['200' is a magic number.](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541123)

File: app/src/main/java/org/wikipedia/imagesearch/Encoder.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+            e.printStackTrace();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+
+        return encodedFile;
+    }
+
+    //Retrieves the file path given a uri
+    private String getPath(Context context, Uri uri ) {
```
Comment (amawai) [inline app/src/main/java/org/wikipedia/imagesearch/Encoder.java]: Codacy Issue found: [')' is preceded with whitespace.](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541123)

File: app/src/main/java/org/wikipedia/main/MainFragment.java [hunk 5]
```diff
@@ -480,6 +617,12 @@ private boolean lastPageViewedWithin(int days) {
         return TimeUnit.MILLISECONDS.toDays(System.currentTimeMillis() - Prefs.pageLastShown()) < days;
     }
 
+    private List <String> getLabelList(String imagePath) throws Exception{
```
Comment (amawai) [inline app/src/main/java/org/wikipedia/main/MainFragment.java]: Codacy Issue found: ['<' is preceded with whitespace.](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541123)

File: app/src/main/java/org/wikipedia/imagesearch/ImageLabeler.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
+package org.wikipedia.imagesearch;
+
+import android.os.AsyncTask;
+import android.util.Base64;
```
Comment (amawai) [inline app/src/main/java/org/wikipedia/imagesearch/ImageLabeler.java]: Codacy Issue found: [Unused import - android.util.Base64.](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541123)

File: app/src/androidTest/java/org/wikipedia/espresso/imagesearch/ImageSearchTest.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+import static android.support.test.espresso.Espresso.onView;
+import static android.support.test.espresso.Espresso.openActionBarOverflowOrOptionsMenu;
+import static android.support.test.espresso.action.ViewActions.click;
+import static android.support.test.espresso.action.ViewActions.closeSoftKeyboard;
+import static android.support.test.espresso.action.ViewActions.replaceText;
+import static android.support.test.espresso.assertion.ViewAssertions.matches;
+import static android.support.test.espresso.matcher.RootMatchers.isDialog;
+import static android.support.test.espresso.matcher.ViewMatchers.isDisplayed;
+import static android.support.test.espresso.matcher.ViewMatchers.withId;
+import static android.support.test.espresso.matcher.ViewMatchers.withText;
```
Comment (amawai) [inline app/src/androidTest/java/org/wikipedia/espresso/imagesearch/ImageSearchTest.java]: Codacy Issue found: [Unused import - android.support.test.espresso.matcher.ViewMatchers.withText.](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541123)

File: app/src/main/java/org/wikipedia/imagesearch/ImageLabeler.java [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+                    sb.append(line + "\n");
+                }
+                response = sb.toString();
+                JSONObject jsonObject = new JSONObject(response);
+
+                JSONArray outputs = jsonObject.getJSONArray("outputs");
+                JSONArray concepts = outputs.getJSONObject(0).getJSONObject("data").getJSONArray("concepts");
+
+                for (int i = 0; i < concepts.length(); i++) {//parses through json for concepts with high correlation
+                    if(i==5){break;}//max 5 results
```
Comment (amawai) [inline app/src/main/java/org/wikipedia/imagesearch/ImageLabeler.java]: Codacy Issue found: ['5' is a magic number.](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541123)

File: app/src/androidTest/java/org/wikipedia/espresso/imagesearch/ImageSearchTest.java [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+import static junit.framework.Assert.assertFalse;
+import static org.hamcrest.Matchers.allOf;
+import static org.hamcrest.Matchers.anything;
```
Comment (amawai) [inline app/src/androidTest/java/org/wikipedia/espresso/imagesearch/ImageSearchTest.java]: Codacy Issue found: [Unused import - org.hamcrest.Matchers.anything.](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541123)

File: app/src/androidTest/java/org/wikipedia/espresso/imagesearch/ImageSearchTest.java [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+import org.junit.Rule;
+import org.junit.Test;
+import org.wikipedia.R;
+import org.wikipedia.main.MainActivity;
+import org.wikipedia.main.MainFragment;
+
+import static android.support.test.InstrumentationRegistry.getInstrumentation;
+import static android.support.test.espresso.Espresso.onData;
+import static android.support.test.espresso.Espresso.onView;
+import static android.support.test.espresso.Espresso.openActionBarOverflowOrOptionsMenu;
```
Comment (amawai) [inline app/src/androidTest/java/org/wikipedia/espresso/imagesearch/ImageSearchTest.java]: Codacy Issue found: [Avoid unused imports such as 'android.support.test.espresso.Espresso.openActionBarOverflowOrOptionsMenu'](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541123)

File: app/src/androidTest/java/org/wikipedia/espresso/imagesearch/ImageSearchTest.java [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+        assertTrue(cameraButton.exists());
+
+        UiObject galleryButton = mDevice.findObject(new UiSelector()
+                .instance(2));
+
+        // test that the gallery button exists
+        assertTrue(galleryButton.exists());
+
+        UiObject photoButton = mDevice.findObject(new UiSelector()
+                .instance(3));
```
Comment (amawai) [inline app/src/androidTest/java/org/wikipedia/espresso/imagesearch/ImageSearchTest.java]: Codacy Issue found: ['3' is a magic number.](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541123)

File: app/src/main/java/org/wikipedia/main/MainFragment.java [hunk 12]
```diff
@@ -159,11 +193,59 @@ public void onActivityResult(int requestCode, int resultCode, final Intent data)
...
+        Toast.makeText(getActivity(), "Recognizing image...", Toast.LENGTH_SHORT).show();
+        onImageSearchProgressBar(true);
+        String encodedImage = imageEncoder.encodeUriToBase64Binary(getContext(), selectedImageUri);
+        updateImageLabels(getContext(), encodedImage);
+    }
+
+    private void updateImageLabels(Context context, String encodedImage) {
+        CallbackTask.execute(() -> getLabelList(encodedImage), new CallbackTask.DefaultCallback<List <String>>() {
+            @Override
+            public void success(List <String> list) {
```
Comment (amawai) [inline app/src/main/java/org/wikipedia/main/MainFragment.java]: Codacy Issue found: ['<' is preceded with whitespace.](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541123)

File: app/src/main/java/org/wikipedia/imagesearch/ImageLabeler.java [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+                    JSONObject concept = concepts.getJSONObject(i);
+                    double value = concept.getDouble("value");
+                    String name = concept.getString("name");
+                    if (value>0.9){
```
Comment (amawai) [inline app/src/main/java/org/wikipedia/imagesearch/ImageLabeler.java]: Codacy Issue found: ['0.9' is a magic number.](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541123)

File: app/src/main/java/org/wikipedia/imagesearch/ImageLabeler.java (hunk 8 above)
Comment (amawai) [inline app/src/main/java/org/wikipedia/imagesearch/ImageLabeler.java]: Codacy Issue found: [';' is not followed by whitespace.](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541123)

File: app/src/androidTest/java/org/wikipedia/espresso/imagesearch/ImageSearchTest.java [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+import static android.support.test.espresso.action.ViewActions.replaceText;
+import static android.support.test.espresso.assertion.ViewAssertions.matches;
+import static android.support.test.espresso.matcher.RootMatchers.isDialog;
+import static android.support.test.espresso.matcher.ViewMatchers.isDisplayed;
+import static android.support.test.espresso.matcher.ViewMatchers.withId;
+import static android.support.test.espresso.matcher.ViewMatchers.withText;
+import static junit.framework.Assert.assertFalse;
+import static org.hamcrest.Matchers.allOf;
+import static org.hamcrest.Matchers.anything;
+import static junit.framework.Assert.assertTrue;
```
Comment (amawai) [inline app/src/androidTest/java/org/wikipedia/espresso/imagesearch/ImageSearchTest.java]: Codacy Issue found: [Wrong order for 'junit.framework.Assert.assertTrue' import.](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541123)

File: app/src/androidTest/java/org/wikipedia/espresso/imagesearch/ImageSearchTest.java [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+import android.support.test.InstrumentationRegistry;
+import android.support.test.espresso.DataInteraction;
+import android.support.test.espresso.ViewInteraction;
+import android.support.test.rule.ActivityTestRule;
+import android.support.test.runner.AndroidJUnit4;
+import android.support.test.uiautomator.UiDevice;
+import android.support.test.uiautomator.UiObject;
+import android.support.test.uiautomator.UiObjectNotFoundException;
+import android.support.test.uiautomator.UiSelector;
+import android.view.View;
```
Comment (amawai) [inline app/src/androidTest/java/org/wikipedia/espresso/imagesearch/ImageSearchTest.java]: Codacy Issue found: [Unused import - android.view.View.](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541123)

File: app/src/main/java/org/wikipedia/feed/searchbar/SearchCardView.java (hunk 1 above)
Comment (amawai) [inline app/src/main/java/org/wikipedia/feed/searchbar/SearchCardView.java]: Codacy Issue found: [Avoid unused imports such as 'android.widget.Toast'](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541123)

File: app/src/main/java/org/wikipedia/main/MainFragment.java [hunk 16]
```diff
@@ -512,10 +655,47 @@ private void closeSearchFragment(@NonNull SearchFragment fragment) {
...
+    }
+
+    @SuppressLint("CommitTransaction")
+    public void closeImageSearchFragment(@NonNull ImageSearchFragment fragment) {
+        getChildFragmentManager()
+                .beginTransaction().remove(fragment).commitNowAllowingStateLoss();
+    }
+
+    @SuppressLint("CommitTransaction")
+    private void openImageSearchFragment(List <String> labels) {
```
Comment (amawai) [inline app/src/main/java/org/wikipedia/main/MainFragment.java]: Codacy Issue found: ['<' is preceded with whitespace.](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541123)

File: app/src/main/java/org/wikipedia/imagesearch/ImageLabeler.java [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+import java.net.HttpURLConnection;
+import java.net.URL;
+
+/**
+ * Created by mnhn3 on 2018-04-09.
+ */
+
+public class ImageLabeler extends AsyncTask<String, Void, String> {
+    private final String KEY;
+    private final String PATH;
```
Comment (amawai) [inline app/src/main/java/org/wikipedia/imagesearch/ImageLabeler.java]: Codacy Issue found: [Name 'PATH' must match pattern '^[a-z][a-zA-Z0-9]*$'.](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541123)

File: app/src/androidTest/java/org/wikipedia/espresso/imagesearch/ImageSearchTest.java [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+import android.view.ViewParent;
+
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.wikipedia.R;
+import org.wikipedia.main.MainActivity;
+import org.wikipedia.main.MainFragment;
+
+import static android.support.test.InstrumentationRegistry.getInstrumentation;
```
Comment (amawai) [inline app/src/androidTest/java/org/wikipedia/espresso/imagesearch/ImageSearchTest.java]: Codacy Issue found: [Avoid unused imports such as 'android.support.test.InstrumentationRegistry.getInstrumentation'](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541123)

File: app/src/androidTest/java/org/wikipedia/espresso/imagesearch/ImageSearchTest.java (hunk 10 above)
Comment (amawai) [inline app/src/androidTest/java/org/wikipedia/espresso/imagesearch/ImageSearchTest.java]: Codacy Issue found: [Unused import - android.support.test.espresso.Espresso.openActionBarOverflowOrOptionsMenu.](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541123)

File: app/src/androidTest/java/org/wikipedia/espresso/imagesearch/ImageSearchTest.java (hunk 9 above)
Comment (amawai) [inline app/src/androidTest/java/org/wikipedia/espresso/imagesearch/ImageSearchTest.java]: Codacy Issue found: [Avoid unused imports such as 'org.hamcrest.Matchers.anything'](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541123)

File: app/src/androidTest/java/org/wikipedia/espresso/imagesearch/ImageSearchTest.java [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+import static android.support.test.espresso.Espresso.onData;
```
Comment (amawai) [inline app/src/androidTest/java/org/wikipedia/espresso/imagesearch/ImageSearchTest.java]: Codacy Issue found: [Unused import - android.support.test.espresso.Espresso.onData.](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541123)

File: app/src/androidTest/java/org/wikipedia/espresso/imagesearch/ImageSearchTest.java [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+import static android.support.test.espresso.Espresso.onView;
+import static android.support.test.espresso.Espresso.openActionBarOverflowOrOptionsMenu;
+import static android.support.test.espresso.action.ViewActions.click;
+import static android.support.test.espresso.action.ViewActions.closeSoftKeyboard;
```
Comment (amawai) [inline app/src/androidTest/java/org/wikipedia/espresso/imagesearch/ImageSearchTest.java]: Codacy Issue found: [Unused import - android.support.test.espresso.action.ViewActions.closeSoftKeyboard.](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541123)

File: app/src/main/java/org/wikipedia/main/MainFragment.java [hunk 21]
```diff
@@ -63,22 +77,33 @@
...
+        LinkPreviewDialog.Callback, ImageSearchFragment.Callback {
     @BindView(R.id.fragment_main_view_pager) ViewPager viewPager;
     @BindView(R.id.fragment_main_nav_tab_layout) NavTabLayout tabLayout;
+    @BindView(R.id.imagesearch_progress) ProgressBar progressBar;
     private Unbinder unbinder;
     private ExclusiveBottomSheetPresenter bottomSheetPresenter = new ExclusiveBottomSheetPresenter();
     private MediaDownloadReceiver downloadReceiver = new MediaDownloadReceiver();
     private MediaDownloadReceiverCallback downloadReceiverCallback = new MediaDownloadReceiverCallback();
+    private Uri outputFileUri;
+    private List <String> imageLabels;
```
Comment (amawai) [inline app/src/main/java/org/wikipedia/main/MainFragment.java]: Codacy Issue found: [Perhaps 'imageLabels' could be replaced by a local variable.](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541123)

File: app/src/test/java/org/wikipedia/imagesearch/EncoderTest.java [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+import static org.mockito.Matchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+/**
+ * Created by amawai on 2018-04-13.
+ */
+
+public class EncoderTest {
```
Comment (abhandal) [inline app/src/test/java/org/wikipedia/imagesearch/EncoderTest.java]: Is it possible to extend the EncoderTest to test edge cases?

File: app/src/main/java/org/wikipedia/imagesearch/ImageSearchFragment.java [hunk 23]
```diff
@@ -0,0 +1,198 @@
...
+
+import butterknife.BindView;
+import butterknife.ButterKnife;
+import butterknife.Unbinder;
+
+/**
+ * Created by amawai on 10/04/18.
+ */
+
+public class ImageSearchFragment extends Fragment implements BackPressedHandler {
```
Comment (abhandal) [inline app/src/main/java/org/wikipedia/imagesearch/ImageSearchFragment.java]: What's the purpose of implement BackPressedHandler for this class?

File: app/src/main/java/org/wikipedia/imagesearch/Encoder.java [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * Created by mnhn3 on 2018-04-12.
+ */
+
+public class Encoder {
+
+    public String encodeFileToBase64Binary(File file) {
```
Comment (abhandal) [inline app/src/main/java/org/wikipedia/imagesearch/Encoder.java]: What's the purpose of this method?

File: app/src/main/java/org/wikipedia/imagesearch/ImageSearchFragment.java (hunk 23 above)
Comment (amawai) [reply app/src/main/java/org/wikipedia/imagesearch/ImageSearchFragment.java]: Since the `ImageSearchFragment` is opened from the `MainFragment`, it replaces the screen of the main feed. Back pressing while in the `ImageSearchFragment` without the `BackPressedHandler`, results in being unable to return to the main feed, as it doesn't get destroyed.

File: app/src/main/java/org/wikipedia/imagesearch/Encoder.java (hunk 24 above)
Comment (amawai) [reply app/src/main/java/org/wikipedia/imagesearch/Encoder.java]: Although this method isn't explicitly used, it could be useful in the future as it allows for various input types if `Context` and `Uri` aren't available, for whatever reason.

File: app/src/main/java/org/wikipedia/imagesearch/Encoder.java [hunk 25]
```diff
@@ -0,0 +1,73 @@
...
+            e.printStackTrace();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+
+        return encodedFile;
+    }
+
+    //Encodes a file given a content Uri as input
+    public String encodeUriToBase64Binary(Context context, Uri uri) {
```
Comment (SunXP) [inline app/src/main/java/org/wikipedia/imagesearch/Encoder.java]: Could you elaborate on the purpose of this method? Why do we need to encode?

File: app/src/main/java/org/wikipedia/main/MainFragment.java [hunk 26]
```diff
@@ -512,10 +645,47 @@ private void closeSearchFragment(@NonNull SearchFragment fragment) {
         getChildFragmentManager().beginTransaction().remove(fragment).commitNowAllowingStateLoss();
     }
 
+    public void switchToSearchFragment(ImageSearchFragment fragment,
```
Comment (SunXP) [inline app/src/main/java/org/wikipedia/main/MainFragment.java]: Could you explain when this fragment would be called and clarify how it would work?

Comment (amawai) [reply app/src/main/java/org/wikipedia/main/MainFragment.java]: This fragment is called when the user is using the Advanced Image Search, where selecting a label opens up the `SearchFragment`. Since the `ImageSearchFragment` is a child fragment of `MainFragment`, the `ImageSearchFragment` has to be closed before opening the `SearchFragment`, which is what `switchToSearchFragment` does. When the `ImageSearchFragment` doesn't get closed, the `SearchFragment` doesn't work properly because it thinks that the `ImageSearchFragment` is its parent, and thus some callback functions don't work as expected, or at all.

File: app/src/main/java/org/wikipedia/imagesearch/Encoder.java (hunk 25 above)
Comment (mnhn329) [reply app/src/main/java/org/wikipedia/imagesearch/Encoder.java]: Currently the images are stored as a URI object, but the API request requires a string that is base64 encoded, so this method provides that conversion

File: README.md [hunk 27]
```diff
@@ -1,7 +1,8 @@
-### Wikipedia Android app
+### Pen Pineapple Apple Pen Wikipedia Android app
```
Comment (jadmalek) [inline README.md]: We've actually already updated the readme to include the codacy badge and a link to our actual wiki, so this will likely conflict when we merge develop onto master.

**Review by SunXP — APPROVED** [review]
The implementation of the image search feature has been well executed with advanced functions that had not been part of the requirements, including the ability to retake a photo if the user is not satisfied with the initial one, the ability to toggle the instant image search or choose from a list of keywords.

**Review by abhandal — APPROVED** [review]
The implementation of the image search is a critical feature that will allow users to enjoy the flexibility of searching articles using images from their gallery and camera. The test coverage within the implementation is significant and covers some edge cases to test the robustness of the system. Additionally, the espresso tests are valid and ensure that UI elements are present and actionable. Overall, a great implementation showcasing even greater collaboration amongst the contributors of the feature.

**Comment by amawai** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
           

Complexity increasing per file
==============================
- app/src/main/java/org/wikipedia/main/MainFragment.java  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/amawai/android-wikipedia-390/pullRequest?prid=1541123&bid=7051231)

File: app/src/test/java/org/wikipedia/imagesearch/EncoderTest.java (hunk 22 above)
Comment (jadmalek) [reply app/src/test/java/org/wikipedia/imagesearch/EncoderTest.java]: Addressed in commit f196158.

**Comment by sophiaquach16** [comment]
**Acceptance Testing**

These acceptance tests characterize the expected behaviour this user story's implementations are to exhibit. The level of accuracy and ability for image search (by camera or alternative image source) to retrieve results/labels based on the indicated images is limited by the API used for the implementation, Clarifai. Also, once an image search is conducted and results are failed to be returned, an error prompt appears and a loading bar for the image search at the top of the main application page will continue going (will not affect functionality of remainder of application).

| Test #         | Inputs | Expected Output  | Status |
| ----------------------- | ------------- | ------------- | ------------- |
| 1. Display of menu to conduct article search by image.  | Select on 'image' button in search bar on main page of the application. | "Image Search: Select Source" appears, along with a list of source options. | PASS  |
| 2. Use of camera for image search.  |  Click on “Camera” option from the menu in test 1.  | Camera usage is now available, along with 'checkmark', 'refresh' and 'X' icons. | PASS
| 3. The photo chosen by the user is being processed.  |  Take a picture of an object or person and select on 'checkmark' icon. | “Recognizing image…” message is displayed. | PASS  |
| 4. Relevant results corresponding to the photo chosen are displayed.  | None. | A list of text card results appear.   | PASS  |
| 5. Display of list of word results correspond to Wikipedia articles.  | Click on a word result from the list. | The Wikipedia article corresponding to that word is displayed to the screen.   | PASS  |
| 6. Image search by selection of a difference image source.  | Repeat input from test 1, select on an alternative image source ('Gallery', for example), select the desired image, and repeat inputs from tests 4 and 5. | Same expected output as test 5.  | PASS

**Demo Steps**

1. Launch the application.
2. Click on the 'image' button of the “Search Wikipedia” bar, the button of a photo right before the microphone button.
3. Click on “Camera” OR “Gallery” for the Image Search: Select Source. 
4. If “Camera is chosen”, take a photo and press the checkmark button to confirm. If “Gallery”, choose the photo to be used.
5. Click on a word result from the list for the article to open up.
